src/types/middleware.ts

```diff
@@ -2,6 +2,7 @@ import InstantSearch from '../lib/InstantSearch';
 import { UiState } from './widget';
 
 export type MiddlewareDefinition = {
+  $$type: string;
```

Comment:
Since users are meant to create their own middleware, I think `$$type` should be optional.